libbeat/cmd/instance/beat.go

```diff
@@ -984,6 +985,64 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg config.Namespace) (outpu
 		return outputs.Group{}, nil
 	}
 
+	// TODO (Tiago): find a proper name for this logger
+	logger := logp.L().Named("ssl.cert.reloader")
+	// Here the output is created and we have acces to the Beat struct (with the manager)
+	// as a workaround we can unpack the new settings and trigger the reload-watcher from here
+
+	// We get an output config, so we extract the 'SSL' bit from it
+	rawTLSCfg, err := cfg.Config().Child("ssl", -1)
+	if err != nil {
+		return outputs.Group{}, fmt.Errorf("could not extract the 'ssl' section of the output config: %w", err)
+	}
+
+	tlsCfg := tlscommon.Config{}
+	if err := rawTLSCfg.Unpack(&tlsCfg); err != nil {
+		logger.Errorf("could not unpack config: %s", err)
+		return outputs.Group{}, fmt.Errorf("unpacking 'ssl' config: %w", err)
+	}
+
+	watchers := make([]*cfgfile.GlobWatcher, 0, len(tlsCfg.CAs))
+	for _, ca := range tlsCfg.CAs {
+		if tlscommon.IsPEMString(ca) {
+			// That's an embedded cert, we're only interested in files
+			continue
+		}
+
+		logger.Debugf("watching '%s' for changes", ca)
+		gw := cfgfile.NewGlobWatcher(ca)
+		// Ignore the first scan as it will always return
+		// true for files changed. The output has not been
+		// started yet, so even if the files have changed since
+		// the Beat started, they don't need to be reloaded
+		gw.Scan()
+
+		watchers = append(watchers, gw)
+	}
+
+	// Watch for file changes while the Beat is alive
+	go func() {
+		// TODO (Tiago): Make it configurable
+		ticker := time.Tick(3 * time.Second)
+
+		for {
+			<-ticker
+			for _, gw := range watchers {
+				files, changed, err := gw.Scan()
+				if err != nil {
+					logger.Warnf("could not scan certificate files: %s", err.Error())
+				}
+
+				if changed {
+					logger.Infof(
+						"some of the following files have been modified: %v, starting %s shurdown",
+						files, b.Info.Beat)
+					b.Manager.Stop()
```

Comment:
I think we want this behavior behind some kind of flag, or at least a warning of some kind.  As a system admin I wouldn't expect modifying a ca file to halt an application.  It would be fairly normal to pre-provision the files and schedule a restart for later.

Reply:
Yes, definitely. I'll implement the configuration part soon.

For now I was looking more for feedback about the overall approach. I'm not a fan of parsing the config in different places, but that was the only way I found to get it working without a huge refactoring.

Reply:
I'm ok with the approach given how we specify CAs today.  It might be nice to change how we specify CAs, so you can list a directory and then we read in all the certs from that directory.  That way you don't have to use the same name as listed in the current config file.

Reply:
That would be interesting, but might as well be a breaking change :/

To be honest, I don't know what is more common out there: to use a folder for each application or have explicitly paths set.

Because its a central piece of security, I like being more explicit, but it does bring other "issues".